FLEDGE_implementation_overview.md

```diff
@@ -0,0 +1,64 @@
+
+Trusted seller signals fetches are started every 100 milliseconds, when there’s a bid a seller executor is waiting to score.  Once the final bid for an auction has been generated, the final signals fetch is immediately issued.  As with bidders, seller executors can be shared between auctions in the same frame running concurrently, which also allows for for shared seller signals fetches.
+
+Once all bids are scored, the result is immediately returned to the Javascript Promise.
+
+## Phase 3 - Reporting Scripts
+On auction completion, Chrome immediately reuses the seller executor to call reportResult(), and then reloads the winning bidder’s executor to call reportWin(), if it’s not still in memory.  The winning bidder executor is not kept in memory to avoid consuming bidder process quota, and to potentially avoid causing deadlock (e..g, if there are 10 auctions, all keeping different currently top-scoring bidders around, but that still need to load more bidders before they know the final winner, no auction could proceed).
+
+Nothing is reported to the server, and the win is not recorded, until the winning ad is navigated to, or deprecatedURNToURL() is invoked, which may happen while still running reporting scripts.  If the ad loads before a reporting script completes, and tries to send a report that depends on reportWin() / reportResult(), and that function hasn’t completed, the report is queued until the script has finished running.  In general, the run duration of these should not be web visible, though if the frame the auction ran in is destroyed before they complete, any currently in-progress work will be aborted.  This is considered a bug that needs to be fixed.  Running reporting scripts can affect the performance of running auctions, since worklet processes used by reporting count towards the global Protect Audience process quotas.
+
+
+## Component Auctions
+In the case of top-level auction with multiple nested component auctions, the auctions are largely run independently, except that the interest group loading phases for all auctions must complete before any auction enters the generate bid phase, and the top-level auction only requests an executor/process for itself once all component auctions have received one, to avoid deadlock.  The top-level auction scores bids as soon as it receives winning bids from each component auction, and has its executor loaded.  Each component auction unloads the seller executor once all bids have been scored, again to avoid deadlock.  So the reporting phase may have to reload both the winning component seller worklet, and the winning bidder worklet.
```

Comment:
>the auctions are largely run independently, except that the interest group loading phases for all auctions must complete

The simple auction Phase description says..

>Interest groups are loaded on a per-buyer basis,

So Chrome will repeatedly execute (pseudo)

"select * from (interest_groups/k_anon/bid_history/win_history) where owner = 'https://igownerorigin.com' 

for that owner in all of the component auctions?

Reply:
Yes.  Hopefully SQLite will cache the results in memory, so we aren't hitting disk, but currently, we will indeed run the entire set of queries multiple times if multiple component auctions share a buyer.  It will do the same if multiple auctions are started at once with the same buyers as well.

Reply:
Looks like we actually have a caching layer added in the last month, which should handle this.  Not familiar with the details.

Reply:
Added mention of the cache - if we're mentioning the database, seems worth mentioning the cache as well.